Added Autogeneration script for generating Voltron.

TimMcJilton opened the pull request:
Two initial currated list of Voltron and VoltronController.

TimMcJilton commented:
Scripts added for #18

TimMcJilton commented:
For now I just did voltron and voltron_controller. But it would be dead simple to add voltron_choosy and voltron_choosy_controller etc.

TimMcJilton commented on voltron_list.yaml:
```diff
@@ -0,0 +1,70 @@
+        
+outputFiles:
+    voltron:
+        <<: *voltron
```
Would you rather this be voltron_mkb that way voltron.txt can be your manually curated list? And later on you can decide whether or not you want to maintain classic voltron.txt vs auto-gen?

48klocs replied:
Hmm, not sure yet. Ideally, I guess it should be scripted?

There looks like a lot of churn happened, so I'll have to review the file load order in the YAML to see if I can cut down on that.

TimMcJilton replied:
Put a comment later on in this about the churn. You can reduce the churn locally for doing a manual diff by doing `cat voltron_mkb.txt| sort -u  | grep -v '^//' > voltronMkbSorted` Which will sort all of the lines and then filter out comments (since comments in voltron and the source may differ and it doesn't matter). From there you can do a diff and then do a wc -l .

```
-[11:04:06][addAutoGenVoltron]$ cat voltron_mkb.txt| sort -u  | grep -v '^//' > voltronMkbSorted                                     ~/code/dim-wish-list-sources
-[11:04:10][addAutoGenVoltron?]$ cat voltron.txt| sort -u  | grep -v '^//' > voltronSorted                                           ~/code/dim-wish-list-sources
-[11:04:13][addAutoGenVoltron?]$ diff voltronSorted voltronMkbSorted|grep dimwishlist | wc -l                                        ~/code/dim-wish-list-sources
1041
```

48klocs commented:
Thanks for this - it'll take me a bit to review, because I'll need to get a Python runtime up on a box first.

As a minor one, `voltron` no longer contains trash rolls (there's `voltron_choosy` which adds them).

TimMcJilton commented:
> Thanks for this - it'll take me a bit to review, because I'll need to get a Python runtime up on a box first.
> 
> As a minor one, `voltron` no longer contains trash rolls (there's `voltron_choosy` which adds them).
I actually didn't include Trash-rolls in it.  So I can do:
`voltron.txt` `voltron_controller.txt` `voltron_choosy.txt` and `voltron_controller_choosy.txt` 

Or instead of `voltron` and `voltron_choosy` I could do `voltron_mkb` and `voltron_mkb_choosy`
```
        fileGroups:
            - pandaPvE
            - pandaGeneric
            - mercules
            - pandaPvP
            - pandaMKB
```

TimMcJilton commented:
So poking a little bit. It appears that there are some random controller lines mixed in with Voltron that I am stripping out for mkb. 

Went ahead and ran this against both voltron and voltron mkb.
`cat voltron_mkb.txt| sort -u  | grep -v '^//' > voltronMkbSorted` 

So it is sorted,t hen ran a diff between voltronSorted and voltronMkbSorted. 

There also seems to be some areas in voltron that aren't in other files. Example:

```
  8 // armor from u/HavocsCall
  7 // https://github.com/HavocsCall/DIMArmorSets/blob/master/Armor%20Wish%20List.txt
  6 
  5 //-----Helmet-----
  4 
  3 //Sniper Reserves and Remote Connection, Precision Weapon Targeting, Sniper Targeting, or Enhanced Sniper Targeting
  2 dimwishlist:item=45&perks=1983861079,1180144002
  1 dimwishlist:item=45&perks=2414923993,1180144002
  0 dimwishlist:item=45&perks=395489854,1180144002
  1 dimwishlist:item=45&perks=2196878528,1180144002
  2 
  3 //Fusion Rifle Reserves and Light Reactor, Fusion Rifle Targeting, or Scatter Projectile Targeting
  4 dimwishlist:item=45&perks=2138431565,4034210761
  5 dimwishlist:item=45&perks=514613051,4034210761
  6 dimwishlist:item=45&perks=3813989326,4034210761
  7 
  8 //Linear Fusion Rifle Reserves and Precision Weapon Targeting, Linear Fusion Targeting or Enhanced Linear Fusion Targeting
  9 dimwishlist:item=45&perks=2073167962,4007230256
 10 dimwishlist:item=45&perks=2414923993,4007230256
 11 dimwishlist:item=45&perks=3621792645,4007230256
 12 
 13 //Heavy Ammo Finder and Ashes to Assets, Enhanced Ashes to Assets, Heavy Lifting, Enhanced Heavy Lifting, or Hands-On
 14 dimwishlist:item=45&perks=1107740237,2867719094
 15 dimwishlist:item=45&perks=2042848612,2867719094
 16 dimwishlist:item=45&perks=1567314459,2867719094
 17 dimwishlist:item=45&perks=2910842650,2867719094
 18 dimwishlist:item=45&perks=433323503,2867719094
 19 
 20 //Grenade Launcher Reserves and Ashes to Assets, Enhanced Ashes to Assets, Heavy Lifting, Enhanced Heavy Lifting, or Hands-On
 21 dimwishlist:item=45&perks=1107740237,330650351
 22 dimwishlist:item=45&perks=2042848612,330650351
 23 dimwishlist:item=45&perks=1567314459,330650351
```

48klocs commented:
OK, so.

I'm running `Python 3.8.1`.

I ran `pip install pyyaml` (to get the `import yaml` piece working).

When I run `python .\bin\generate_voltrons.py`, I get the following output.

`  File ".\bin\generate_voltrons.py", line 21
    print "Generating {}.txt".format(k)
          ^
SyntaxError: invalid syntax`

TimMcJilton commented:
> OK, so.
> 
> I'm running `Python 3.8.1`.
> 
> I ran `pip install pyyaml` (to get the `import yaml` piece working).
> 
> When I run `python .\bin\generate_voltrons.py`, I get the following output.
> 
> ```
> File ".\bin\generate_voltrons.py", line 21
>     print "Generating {}.txt".format(k)
>         ^
> SyntaxError: invalid syntax
> ```
> 
> I also touched the source YAML file and added an `originalVoltron` group that should regenerate everything (I also added a new `armor_1_picks.txt` in the upstream master.

Yeah... you .. you can judge me for using python 2.7 still.. I never moved forward with the rest of society yet. Easy fix though.

TimMcJilton commented:
It should be good now. I didn't want to deal with having two sets of python installed locally (even though I know pyenv exists for that). So I  futurized it. in _theory_ it should work fine for python 3.8 now. Works on my python2.7 now.

48klocs commented:
λ python .\bin\generate_voltrons.py
.\bin\generate_voltrons.py:10: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  ymlRaw = yaml.load(f)
Generating voltron_mkb.txt
Traceback (most recent call last):
  File ".\bin\generate_voltrons.py", line 73, in <module>
    main()
  File ".\bin\generate_voltrons.py", line 24, in main
    v.printFile()
  File ".\bin\generate_voltrons.py", line 41, in printFile
    f.write(self.getString())
  File ".\bin\generate_voltrons.py", line 38, in getString
    return "\n".join(self.getLines())
  File ".\bin\generate_voltrons.py", line 35, in getLines
    lines = lines + [""] + f.getLines()
  File ".\bin\generate_voltrons.py", line 53, in getLines
    lines = lines + file.readFile()
  File ".\bin\generate_voltrons.py", line 67, in readFile
    lines = f.readlines()
  File "D:\Program Files\Python\Python38\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 1563: character maps to <undefined>

TimMcJilton commented:
I am traveling until next Saturday, can you send me what version of python and what version of pyaml installed? Next time I have the chance I will spin up a docker container and see if I can reproduce these errors. But it won't be until the 27th.(may be worth getting on discord and trying to get it working on your box )

48klocs commented:
My local Python install is `3.8.1`, PyYAML is `5.3`.

I'm OK with waiting to get this merged in (and am more than happy to chat on Discord when you're back and we have time rather than playing tag in the comments here).

I'm fine with documenting install/run steps for this (`pyenv`, which I don't have installed yet, seems like a no-brainer if it's targeting an older version), but I'd kind of like to get it to a place where I can run it on a couple of machines before including it for more general consumption, since I'd like to be able to respond meaningfully to any support requests that crop up.

    λ python --version
    Python 3.8.1

    λ pip show pyyaml
    Name: PyYAML
    Version: 5.3

TimMcJilton commented:
> My local Python install is `3.8.1`, PyYAML is `5.3`.
> 
> I'm OK with waiting to get this merged in (and am more than happy to chat on Discord when you're back and we have time rather than playing tag in the comments here).
> 
> I'm fine with documenting install/run steps for this (`pyenv`, which I don't have installed yet, seems like a no-brainer if it's targeting an older version), but I'd kind of like to get it to a place where I can run it on a couple of machines before including it for more general consumption, since I'd like to be able to respond meaningfully to any support requests that crop up.
> 
> ```
> λ python --version
> Python 3.8.1
> 
> λ pip show pyyaml
> Name: PyYAML
> Version: 5.3
> ```

Looking into this now. I'm going to try to get it working with your setup vs mine since newest pyyaml and python is probably better than python 2 and pyyaml 3.13.  Sorry was traveling then came home sick for a week. Sometime this week I'll find time to fix it. Though.. your error makes me think it's an issue with a codec ```File "D:\Program Files\Python\Python38\lib\encodings\cp1252.py", line 23, in decode
return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 1563: character maps to <undefined>``` Like.. it's not a big fan of your pukes.

48klocs commented:
Peeked at latest version, same error running locally.

```
  File "C:\Users\dave.solomon\Documents\GitHub\dim-wish-list-sources\bin\generate_voltrons.py", line 21
    print "Generating {}.txt".format(k)
          ^
SyntaxError: invalid syntax
```

TimMcJilton commented:
Not sure what the next step is for this. Worked on 2 of my ubuntu machines + 1 Windows machine. I can rewrite this in go in a pretty short amount of time if you want me to just upload a compiled binary you can run. It's pretty simple code-wise and would be dead simple to translate.

tuzonghua commented:
Just a rando jumping in, but I'm happy to test/debug this more if necessary. @48klocs your error suggests you aren't running the latest version since that line is wrapped in parens to make it Py3 friendly (see [here](https://github.com/TimMcJilton/dim-wish-list-sources/blob/addAutoGenVoltron/bin/generate_voltrons.py#L24)).

TimMcJilton commented:
> Just a rando jumping in, but I'm happy to test/debug this more if necessary. @48klocs your error suggests you aren't running the latest version since that line is wrapped in parens to make it Py3 friendly (see [here](https://github.com/TimMcJilton/dim-wish-list-sources/blob/addAutoGenVoltron/bin/generate_voltrons.py#L24)).

Ah, I should have updated with some of the conversation offline. @48klocs re-downloaded teh source and tried out the code. His error ended up becoming
```
λ python .\bin\generate_voltrons.py
Traceback (most recent call last):
  File ".\bin\generate_voltrons.py", line 74, in <module>
    main()
  File ".\bin\generate_voltrons.py", line 21, in main
    outputFiles[k] = OutputFile(k, v, fileGroups)
  File ".\bin\generate_voltrons.py", line 30, in __init__
    self.description = fileDefinition["description"]
KeyError: 'description'
```
Which perplexes me. I asked him to print out the yaml he loaded and he got 
```
{'genericVoltron': {'title': 'This is a compiled collection of god/recommended rolls from top community minds.', 'description': 'Thanks to u/Mercules904, u/pandapaxxy, u/HavocsCall, and @chrisfried for their contributions. Ping u/48klocs / @48klocs with any issues.'}, 'definitions': {'originalVoltron': ['mercules_erentil_and_opulence_iron_banner.txt', 'mercules_menagerie.txt', 'mercules_crown_of_sorrow.txt', 'panda_penumbra_pvp.txt', 'panda_penumbra_pve.txt', 'mercules_gambit_rolls.txt', 'mercules_iron_banner.txt', 'mercules_iron_banner_bonus.txt', 'mercules_last_wish_rolls.txt', 'mercules_black_armory_and_sotp_rolls.txt', 'mercules_black_armory_pulse_bow.txt', 'panda_black_armory_picks.txt', 'panda_missing_black_armory_pve.txt', 'panda_missing_black_armory_pvp.txt', 'panda_season_drifter_pvp.txt', 'panda_season_drifter_pve.txt', 'armor_1_picks.txt', 'panda_shadowkeep_pve.txt', 'mercules_vex_offensive_breakdown.txt', 'mercules_undying_season_pass_picks.txt', 'mercules_altar_of_sorrow.txt', 'panda_pit_and_altar.txt', 'panda_dawn_ritual.txt', 'panda_sundial_mkb.txt', 'panda_sundial_controller.txt', 'panda_dawning_mkb.txt', 'mercules_season_of_dawn_season_pass_breakdown.txt', 'panda_dawning_controller.txt', 'mercules_osiris_redux.txt', 'panda_returning_dawn_mkb.txt', 'panda_returning_dawn_controller.txt', 'mercules_garden_of_salvation.txt', 'panda_salvation_pve.txt', 'panda_salvation_pvp.txt'], 'pandaPvP': ['panda_black_armory_pvp_picks.txt', 'panda_missing_black_armory_pvp.txt', 'panda_salvation_pvp.txt', 'panda_shadowkeep_pvp.txt', 'panda_forsaken_pvp_rolls.txt', 'panda_penumbra_pvp.txt', 'panda_season_drifter_pvp.txt'], 'pandaPvE': ['panda_black_armory_pve_picks.txt', 'panda_missing_black_armory_pve.txt', 'panda_forsaken_pve_rolls.txt', 'panda_salvation_pve.txt', 'panda_season_drifter_pve.txt', 'panda_shadowkeep_pve.txt'], 'pandaGeneric': ['panda_black_armory_picks.txt', 'panda_dawn_ritual.txt', 'panda_penumbra_pve.txt', 'panda_pit_and_altar.txt'], 'pandaMKB': ['panda_returning_dawn_mkb.txt', 'panda_dawning_mkb.txt', 'panda_sundial_mkb.txt'], 'pandaController': ['panda_returning_dawn_controller.txt', 'panda_dawning_controller.txt', 'panda_sundial_controller.txt'], 'mercules': ['mercules_altar_of_sorrow.txt', 'mercules_black_armory_and_sotp_rolls.txt', 'mercules_black_armory_pulse_bow.txt', 'mercules_crown_of_sorrow.txt', 'mercules_dreaming_city_breakdown_rolls.txt', 'mercules_erentil_and_opulence_iron_banner.txt', 'mercules_gambit_rolls.txt', 'mercules_garden_of_salvation.txt', 'mercules_iron_banner_bonus.txt', 'mercules_iron_banner.txt', 'mercules_last_wish_rolls.txt', 'mercules_menagerie.txt', 'mercules_osiris_redux.txt', 'mercules_season_of_dawn_season_pass_breakdown.txt', 'mercules_undying_season_pass_picks.txt', 'mercules_vex_offensive_breakdown.txt'], 'trash_lists': ['trash_rolls.txt']}, 'outputFiles': {'voltron_mkb': {'title': 'This is a compiled collection of god/recommended rolls from top community minds.', 'description': 'Thanks to u/Mercules904, u/pandapaxxy, u/HavocsCall, and @chrisfried for their contributions. Ping u/48klocs / @48klocs with any issues.', 'fileGroups': ['pandaPvE', 'pandaGeneric', 'mercules', 'pandaPvP', 'pandaMKB']}, 'voltron_controller': {'title': 'This is a compiled collection of god/recommended rolls from top community minds.', 'description': 'Thanks to u/Mercules904, u/pandapaxxy, u/HavocsCall, and @chrisfried for their contributions. Ping u/48klocs / @48klocs with any issues.', 'fileGroups': ['pandaPvE', 'pandaGeneric', 'mercules', 'pandaPvP', 'pandaController']}, 'voltron_mkb_choosy': {'title': 'This is a compiled collection of god/recommended rolls from top community minds.', 'description': 'Thanks to u/Mercules904, u/pandapaxxy, u/HavocsCall, and @chrisfried for their contributions. Ping u/48klocs / @48klocs with any issues.', 'fileGroups': ['pandaPvE', 'pandaGeneric', 'mercules', 'pandaPvP', 'pandaMKB', 'trash_lists']}, 'voltron_controller_choosy': {'title': 'This is a compiled collection of god/recommended rolls from top community minds.', 'description': 'Thanks to u/Mercules904, u/pandapaxxy, u/HavocsCall, and @chrisfried for their contributions. Ping u/48klocs / @48klocs with any issues.', 'fileGroups': ['pandaPvE', 'pandaGeneric', 'mercules', 'pandaPvP', 'pandaController', 'trash_lists']}, 'voltron': {'fileGroups': ['originalVoltron']}, 'choosy_voltron': {'fileGroups': ['originalVoltron', 'trash_lists']}}}
```

His error definitely confuses me.. because if he couldn't load the yaml it would make sense, but he was able to.. But then I was thinking "maybe he didn't have the collections library..." (which arguably could be a version issue there?)

It'd be bad to have a release-strategy that involved @48klocs relying on a 3-rd party to regen.

48klocs commented:
If there are other libraries that I should have installed locally, I almost certainly don't.

There has to be some sort of way to declare dependencies, no?

TimMcJilton commented:
So.. That is one perk JS + Node has over python the idea of creating a quick lil virtual environment to install dependencies and such into. The best Python has for dependency manager are either making the python a package with a setup.py, or just a plain txt file with the list of files and versions.

The part that I can't wrap my mind against is I only installed very very basic packages, based on what you told me you had installed.
```
PS C:\Users\micro> python -m pip list
DEPRECATION: The default format will switch to columns in the future. You can use --format=(legacy|columns) (or define a format=(legacy|columns) in your pip.conf under the [list] section) to disable this warning.
pip (9.0.3)
PyYAML (5.3)
setuptools (39.0.1)
You are using pip version 9.0.3, however version 20.0.2 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.
```

Like.. 
```
PS C:\Users\micro> python --version
Python 3.6.5
```

I really don't have much going on that is special. I can re-write the script pretty fast in a compile language if you want, there isn't like.. any logic to it. It reads in yaml, and then all the files, saves it in a hashmap, and then prints out the new one. I would offer to do JS except JS and I don't usually get along.

luislobo commented:
On my end, to make it work on python 2.7.15rc1 (ubuntu on windows) I had to change the code like this:
add `import.io`

prepend every `open(` with `io.open(`

TimMcJilton commented:
@48klocs I'm going to go ahead and close this. If you want it or want to figure it out, feel free to keep it. But at this point I am not going to end up fixing since you don't seem interested in it.

Cheers, and thanks for everything you do.
